pageserver/src/tenant/timeline.rs

```diff
@@ -1082,7 +1082,10 @@ impl Timeline {
                 }
 
                 if let Some(layer_size) = layer_size {
-                    self.metrics.resident_physical_size_gauge.sub(layer_size);
+                    let current_size = self.metrics.resident_physical_size_gauge.get();
+                    self.metrics
+                        .resident_physical_size_gauge
+                        .set(current_size.saturating_sub(layer_size));
```

Comment:
good point. 
the gauge itself is atomic, but there is no lock between get and set.

it this version better?
```
if ( self.metrics.resident_physical_size_gauge > layer_size)
{
 self.metrics.resident_physical_size_gauge.sub(layer_size);
}
```

Though, this will fail too, if several jobs can evict layers simultaneously for one timeline...

Reply:
you can extract the gauge update to a method with a local lock but then you will need to do so to add as well

Reply:
I explored the code a bit more and I can't find the race with initialization.
The gauge is initialized before we run eviction. So, the problem must be different.

Here is a [dashboard](https://neonprod.grafana.net/d/QzLjnx14z/metrics-playground-per-tenant?orgId=1&var-data_source=victoria-metrics-aws-prod&var-tenant_id=6927633d6d4a40530b353471bac667de&var-timeline_id=fc9374f3882c816bf97eaf1270af726b&var-tenant_instance=pageserver-0.ap-southeast-1.aws.neon.tech&from=1677517357258&to=1677534023767) that illustrates the problematic moment.
_Note, there is a pageserver log at the bottom of the page and you can conveniently match time range on all panels._

`resident_physical_size` jumped to abnormal value right when `pageserver_remote_physical_size` went down
This moment matches eviction_task completion in the log:
```
2023-02-27T18:07:16.220248Z  INFO eviction_task{tenant_id=6927633d6d4a40530b353471bac667de timeline_id=fc9374f3882c816bf97eaf1270af726b}:eviction_iteration{policy_kind="LayerAccessThreshold"}: Rebuilt layer map. Did 406 insertions to process a batch of 1 updates.
```

What surprises me even more, is that the metric went back to normal (~2.5 GB)  ~after an "empty" GC run:
```
2023-02-27T20:06:26.423326Z  INFO gc_loop{tenant_id=6927633d6d4a40530b353471bac667de}:gc_timeline{timeline=fc9374f3882c816bf97eaf1270af726b cutoff=8B/78F99CE8}: GC completed removing 0 layers, cutoff 8B/78F99CE8
```

My next hypothesis is that eviction and GC conflict somehow and update the same gauge twice (?), which causes overflow.

Can anyone with better understanding of layer eviction check this?